internal/resourcecache/resourcecache.go

```diff
@@ -0,0 +1,109 @@
+	resources map[string]*Resource
+	sync.Mutex
+}
+
+// Resource contains the actual resource itself (which must implement the IdentifiableCreatable interface),
+// as well as stores function pointers that pertain to how that resource should be cleaned up.
+type Resource struct {
+	resource     IdentifiableCreatable
+	cleanupFuncs []func()
+}
+
+// IdentifiableCreatable are the qualities needed by the caller of the resource.
+// Once a resource is retrieved, SetCreated() will be called, indicating to the server
+// that resource is ready to be listed and operated upon, and ID() will be used to identify the
+// newly created resource to the server.
+type IdentifiableCreatable interface {
+	ID() string
+	SetCreated()
+}
+
+// New creates a new ResourceCache
+func New() *ResourceCache {
+	rc := &ResourceCache{
+		resources: make(map[string]*Resource),
+	}
+	return rc
+}
+
+// AddResourceToCache takes a unique resource name (retrieved from the client request, not generated by the server)
+// a newly created resource, and functions to cleanup that newly created resource.
+// It adds the Resource to the ResourceCache, as well as starts a go routine that is responsible for cleaning up the
+// resource if the server has not gotten another request for it (i.e. if ResourceFromCache has not been called).
+// AddResourceToCache expects `name`s passed to it to be unique. If a duplicate `name` is detected, it returns an error.
+func (rc *ResourceCache) AddResourceToCache(name string, resource IdentifiableCreatable, cleanupFuncs []func()) error {
+	return rc.AddResourceToCacheWithTimeout(name, resource, cleanupFuncs, sleepTimeBeforeCleanup)
+}
+
+// AddResourceToCacheWithTimeout is used for testing purposes. It allows the caller to set the timeout, allowing for faster tests.
+// Most callers should use AddResourceToCache instead.
+func (rc *ResourceCache) AddResourceToCacheWithTimeout(name string, resource IdentifiableCreatable, cleanupFuncs []func(), timeout time.Duration) error {
+	rc.Lock()
+	defer rc.Unlock()
+
+	if _, ok := rc.resources[name]; ok {
+		return errors.Errorf("failed to add entry %s to ResourceCache; entry already exists", name)
+	}
+	rc.resources[name] = &Resource{
+		resource:     resource,
+		cleanupFuncs: cleanupFuncs,
+	}
+	go rc.cleanupResourceAfterSleep(name, timeout)
+	return nil
+}
+
+// cleanupResourceAfterSleep sleeps for `sleepTimeBeforeCleanup` and then attempts to
+// remove the entry from the ResourceCache.
+// To remove the entry, the item is removed from the cache, and its cleanupFuncs are called.
+func (rc *ResourceCache) cleanupResourceAfterSleep(name string, timeout time.Duration) {
+	time.Sleep(timeout)
+
+	rc.Lock()
+	r, ok := rc.resources[name]
+	if !ok {
+		rc.Unlock()
+		return
+	}
+	logrus.Infof("cleaning up %s after kubelet didn't re-request it", name)
+	delete(rc.resources, name)
+	// no need to hold the lock when running the cleanup functions
+	rc.Unlock()
+
+	for _, f := range r.cleanupFuncs {
+		f()
+	}
```

Comment:
If the resource is removed and then instantly re-added, this code might call its cleanup functions while it's being re-added, i.e. at least theoretically there is a race here.

Reply:
*maybe* for users that aren't using pod/container name as a way to determine whether there's a duplicated request going on. the last cleanupFunc to be called is `Release{Pod,Container}Name` so, at worse, a goroutine adds itself as a watcher for a resource that's already been cleaned up, but that will time out in 4 minutes. Unless I misunderstand the flow you're imagining